### PR TITLE
Convert AddCommandTest to a system test #551

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Part of #551.

```
Tests in the 'guitests' packages, such as AddCommandTest, merely
concern themselves with the relevant GUI components and the model, and
do not verify other parts of the application (e.g. storage, status bar,
...).

They do not fully qualify as system tests, and we already have
unit tests and integration tests for commands, thus making them
redundant. Also, without any system tests in the codebase, we are unable
to demonstrate to the students how one should be written.

Let's convert AddCommandTest to a system test.
```

### CODE REVIEW

**Comment by Zhiyuan-Amos** [comment]
I should probably add a test case without tags.

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/2] [Convert AddCommandTest to AddCommandSystemTest](https://github.com/se-edu/addressbook-level4/pull/640/commits/aa84ab8cd6fc4d9d40dc8d0897f3f514f588db21)
* [v1 2/2] [AddCommandSystemTest: Add additional tests](https://github.com/se-edu/addressbook-level4/pull/640/commits/781f83713d6705996a73734b39d360183d2448a1)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/e32a1e791badb4acfe80365177b0600826dcfee1...781f83713d6705996a73734b39d360183d2448a1))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/640/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
Made some changes as the current version is too messy.

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/2] [Convert AddCommandTest to AddCommandSystemTest](https://github.com/se-edu/addressbook-level4/pull/640/commits/6a33c86ff5f339c6ac04b764f7702da4cd01bc0c)
* [v2 2/2] [AddCommandSystemTest: Add additional tests](https://github.com/se-edu/addressbook-level4/pull/640/commits/08996dd9547eb1e4778807473b9b0144a8fa08b5)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/e32a1e791badb4acfe80365177b0600826dcfee1...08996dd9547eb1e4778807473b9b0144a8fa08b5)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/6bbb2cba0918384eb703856f10fd550a8cccfa50~1...6bbb2cba0918384eb703856f10fd550a8cccfa50))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/640/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/2] [Convert AddCommandTest to AddCommandSystemTest](https://github.com/se-edu/addressbook-level4/pull/640/commits/6a33c86ff5f339c6ac04b764f7702da4cd01bc0c)
* [v3 2/2] [AddCommandSystemTest: Add additional tests](https://github.com/se-edu/addressbook-level4/pull/640/commits/08996dd9547eb1e4778807473b9b0144a8fa08b5)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/e32a1e791badb4acfe80365177b0600826dcfee1...08996dd9547eb1e4778807473b9b0144a8fa08b5)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/e4f7aa63cc4766377d2a1f772999be66a63b37f9~1...e4f7aa63cc4766377d2a1f772999be66a63b37f9))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/640/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
V.2 and V.3 are the same, accidentally submitted twice ;<

**Comment by Zhiyuan-Amos** [comment]
Note to self: Add missing test: FindCommand then AddCommand

**Comment by Zhiyuan-Amos** [comment]
@yamgent for your review here too :P

**Comment by Zhiyuan-Amos** [comment]
Add a test case with multiple whitespaces between fields.

**Comment by Zhiyuan-Amos** [comment]
Rebased with master.

**Comment by Zhiyuan-Amos** [comment]
@damithc prof, tagging you for review as well for the same reason as `ClearCommandTest` PR #639

File: src/test/java/systemtests/AddCommandSystemTest.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+        /* Case: redo adding Amy to the list -> Amy added again */
+        command = RedoCommand.COMMAND_WORD;
+        model.addPerson(toAdd);
+        expectedResultMessage = RedoCommand.MESSAGE_SUCCESS;
+        assertCommandSuccess(command, model, expectedResultMessage);
+
+        /* Case: add a duplicate person -> rejected */
+        command = PersonUtil.getAddCommand(toAdd);
+        assertCommandFailure(command, AddCommand.MESSAGE_DUPLICATE_PERSON);
```
Comment (damithc) [inline src/test/java/systemtests/AddCommandSystemTest.java]: You may need more test cases to test the definition of 'duplicate'

File: src/test/java/systemtests/AddCommandSystemTest.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        } catch (DuplicatePersonException dpe) {
+            throw new IllegalArgumentException("toAdd already exists in the model.");
+        }
+        String expectedResultMessage = String.format(AddCommand.MESSAGE_SUCCESS, toAdd);
+
+        assertCommandSuccess(command, expectedModel, expectedResultMessage);
+    }
+
+    /**
+     * Performs the same verification as {@code assertCommandSuccess(String, ReadOnlyPerson)} except that the result
```
Comment (damithc) [inline src/test/java/systemtests/AddCommandSystemTest.java]: This sentence starts differently from that of the previous method. Any reason? If not, phrase similarly.

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 3]
```diff
@@ -56,13 +57,17 @@
...
+
     static {
         DESC_AMY = new EditPersonDescriptorBuilder().withName(VALID_NAME_AMY)
                 .withPhone(VALID_PHONE_AMY).withEmail(VALID_EMAIL_AMY).withAddress(VALID_ADDRESS_AMY)
                 .withTags(VALID_TAG_FRIEND).build();
         DESC_BOB = new EditPersonDescriptorBuilder().withName(VALID_NAME_BOB)
                 .withPhone(VALID_PHONE_BOB).withEmail(VALID_EMAIL_BOB).withAddress(VALID_ADDRESS_BOB)
                 .withTags(VALID_TAG_HUSBAND, VALID_TAG_FRIEND).build();
+        AMY = new PersonBuilder().withName(VALID_NAME_AMY).withPhone(VALID_PHONE_AMY).withEmail(VALID_EMAIL_AMY)
+                .withAddress(VALID_ADDRESS_AMY).withTags(VALID_TAG_FRIEND).build();
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Feels a bit weird to put it here since we always put these in `TypicalPersons`.

File: src/test/java/systemtests/AddCommandSystemTest.java [hunk 4]
```diff
@@ -1,47 +1,124 @@
...
+        assertCommandFailure(command, Email.MESSAGE_EMAIL_CONSTRAINTS);
+
+        /* Case: invalid address -> rejected */
+        command = AddCommand.COMMAND_WORD + NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY + INVALID_ADDRESS_DESC;
+        assertCommandFailure(command, Address.MESSAGE_ADDRESS_CONSTRAINTS);
+
+        /* Case: invalid tag -> rejected */
+        command = AddCommand.COMMAND_WORD + NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY + ADDRESS_DESC_AMY
+                + INVALID_TAG_DESC;
+        assertCommandFailure(command, Tag.MESSAGE_TAG_CONSTRAINTS);
```
Comment (yamgent) [inline src/test/java/systemtests/AddCommandSystemTest.java]: Should we also test missing compulsory fields?

Comment (Zhiyuan-Amos) [reply src/test/java/systemtests/AddCommandSystemTest.java]: good catch!

**Comment by Zhiyuan-Amos** [comment]
Take note of [3/3] ^^

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java (hunk 3 above)
Comment (damithc) [reply src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: We can put common data in TypicalPersons and test specific data in those tests themselves.

File: src/test/java/systemtests/AddCommandSystemTest.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        /* Case: add a duplicate person -> rejected */
+        command = AddCommand.COMMAND_WORD + NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY + ADDRESS_DESC_AMY
+                + TAG_DESC_FRIEND;
+        assertCommandFailure(command, AddCommand.MESSAGE_DUPLICATE_PERSON);
+
+        // The below test case fails because of a bug in AddressBook#addPerson(ReadOnlyPerson)
+        /* Case: add a duplicate person except with different tags -> rejected */
+        //command = AddCommand.COMMAND_WORD + NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY + ADDRESS_DESC_AMY
+        //        + TAG_DESC_HUSBAND;
+        //assertCommandFailure(command, AddCommand.MESSAGE_DUPLICATE_PERSON);
```
Comment (damithc) [inline src/test/java/systemtests/AddCommandSystemTest.java]: Did you mean two persons with same data but different tags are currently considered duplicates? It can be considered a feature rather than a bug. i.e. that's how we define 'duplicate'. In practice, detecting duplicate persons should be based on a minimum number of fields (and a less than 100% similarity) rather than a maximum. That's because we add a person for the second time, we are unlikely to enter the details exactly as the previous time.

File: src/test/java/systemtests/AddCommandSystemTest.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+        /* Case: add a duplicate person except with different name -> added */
```
Comment (damithc) [inline src/test/java/systemtests/AddCommandSystemTest.java]: `duplicate person except with different name`: it is a bit confusing to call there as 'duplicates' (because they are not). `all fields same except name`?

File: src/test/java/systemtests/AddCommandSystemTest.java (hunk 5 above)
Comment (Zhiyuan-Amos) [reply src/test/java/systemtests/AddCommandSystemTest.java]: Oh nope. When we add a new `Person` with new `Tag`s i.e. `Tag`s that are not yet in `UniqueTagList`, there will be a bug if the adding of the `Person` fails. This is because the new `Tag` is synchronised with the `UniqueTagList` first, before attempting to add the `Person`. As such, if the adding of the `Person` fails, the `UniqueTagList` is incorrect. Copying the code over for your reference.

```
public void addPerson(ReadOnlyPerson p) throws DuplicatePersonException {
    Person newPerson = new Person(p);
    syncMasterTagListWith(newPerson);
    // TODO: the tags master list will be updated even though the below line fails.
    // This can cause the tags master list to have additional tags that are not tagged to any person
    // in the person list.
    persons.add(newPerson);
}
```
